server/src/main/java/org/elasticsearch/index/codec/bloomfilter/BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,472 @@
+            assert false : "ESBloomFilterPostingsFormat was initialized with a wrong constructor";
+            throw new UnsupportedOperationException("ESBloomFilterPostingsFormat was initialized with a wrong constructor");
+        }
+        return new FieldsWriter(postingsFormat, state);
+    }
+
+    @Override
+    public FieldsProducer fieldsProducer(SegmentReadState state) throws IOException {
+        return new FieldsReader(state);
+    }
+
+    @Override
+    public String toString() {
+        return "ESBloomFilterPostingsFormat(" + postingsFormat.getName() + ")";
+    }
+
+    private static String metaField(SegmentInfo si, String segmentSuffix) {
+        return IndexFileNames.segmentFileName(si.name, segmentSuffix, BLOOM_FILTER_META_FILE);
+    }
+
+    private static String indexField(SegmentInfo si, String segmentSuffix) {
+        return IndexFileNames.segmentFileName(si.name, segmentSuffix, BLOOM_FILTER_INDEX_FILE);
+    }
+
+    final class FieldsWriter extends FieldsConsumer {
+        private final FieldsConsumer termsWriter;
+        private final SegmentWriteState state;
+        private final IndexOutput indexOut;
+        private final List<BloomFilter> bloomFilters = new ArrayList<>();
+        private boolean closed;
+
+        FieldsWriter(PostingsFormat postingsFormat, SegmentWriteState state) throws IOException {
+            this.state = state;
+            final List<Closeable> toCloses = new ArrayList<>(1);
+            try {
+                termsWriter = postingsFormat.fieldsConsumer(state);
+                toCloses.add(termsWriter);
+                indexOut = state.directory.createOutput(indexField(state.segmentInfo, state.segmentSuffix), state.context);
+                toCloses.add(indexOut);
+                CodecUtil.writeIndexHeader(indexOut, BLOOM_CODEC_NAME, VERSION_CURRENT, state.segmentInfo.getId(), state.segmentSuffix);
+                toCloses.clear();
+            } finally {
+                IOUtils.closeWhileHandlingException(toCloses);
+            }
+        }
+
+        @Override
+        public void write(Fields fields, NormsProducer norms) throws IOException {
+            termsWriter.write(fields, norms);
+            for (String field : fields) {
+                final Terms terms = fields.terms(field);
+                if (terms == null) {
+                    continue;
+                }
+                final int bloomFilterSize = bloomFilterSize(state.segmentInfo.maxDoc());
+                final int numBytes = numBytesForBloomFilter(bloomFilterSize);
+                try (ByteArray buffer = bigArrays.newByteArray(numBytes)) {
+                    final TermsEnum termsEnum = terms.iterator();
+                    while (true) {
+                        final BytesRef term = termsEnum.next();
+                        if (term == null) {
+                            break;
+                        }
+                        // Do we need to skip terms without postings?
```

Comment:
It shouldn't be necessary, flush and merge guarantee that all terms would have postings when we get here

Reply:
ok, I removed this TODO